File: examples/www/public/assets/scripts/music.lua
local Music = {}
Music.__index = Music

function Music.load(path)
  local handle = load_music(path)
  if handle == nil or handle == 0 then
    return nil
  end

  return setmetatable({
    handle = handle,
  }, Music)
end

function Music:play()
  if self.handle ~= nil and self.handle ~= 0 then
    play_music(self.handle)
  end
end

function Music:pause()
  if self.handle ~= nil and self.handle ~= 0 then
    pause_music(self.handle)
  end
end

function Music:stop()
  if self.handle ~= nil and self.handle ~= 0 then
    stop_music(self.handle)
  end
end

function Music:is_playing()
  if self.handle == nil or self.handle == 0 then
    return false
  end

  return is_music_playing(self.handle)
end

function Music:set_loop(loop)
  if self.handle ~= nil and self.handle ~= 0 then
    set_music_loop(self.handle, loop)
  end
end

function Music:set_volume(volume)
  if self.handle ~= nil and self.handle ~= 0 then
    set_music_volume(self.handle, volume)
  end
end

function Music:destroy()
  if self.handle ~= nil and self.handle ~= 0 then
    destroy_music(self.handle)
    self.handle = 0
  end
end

return Music

File: examples/remote/server/src/music.ts
import type { ResourceManager } from "./resource_manager";
import {
  rl_music_create,
  rl_music_destroy,
  rl_music_pause,
  rl_music_play,
  rl_music_set_loop,
  rl_music_set_volume,
  rl_music_stop,
  set_resource_manager as set_music_resource_manager,
} from "./rl/rl_music";

type MusicAction = "none" | "play" | "pause" | "stop";

export class Music {
  handle: number | null = null;
  private _loop = false;
  private _volume = 1.0;
  private settingsDirty = false;
  private pendingAction: MusicAction = "none";

  constructor(
    public readonly path: string,
    private readonly resourceManager: ResourceManager,
  ) {}

  static async load(resourceManager: ResourceManager, path: string): Promise<Music> {
    const music = new Music(path, resourceManager);
    set_music_resource_manager(resourceManager);
    music.handle = await rl_music_create(path);
    return music;
  }

  play(): void {
    if (this.handle == null) {
      return;
    }
    this.pendingAction = "play";
  }

  pause(): void {
    if (this.handle == null) {
      return;
    }
    this.pendingAction = "pause";
  }

  stop(): void {
    if (this.handle == null) {
      return;
    }
    this.pendingAction = "stop";
  }

  sync(): boolean {
    if (this.handle == null) {
      return false;
    }

    let changed = false;

    if (this.settingsDirty) {
      changed = rl_music_set_loop(this.handle, this._loop) || changed;
      changed = rl_music_set_volume(this.handle, this._volume) || changed;
      this.settingsDirty = false;
    }

    switch (this.pendingAction) {
      case "play":
        changed = rl_music_play(this.handle) || changed;
        break;
      case "pause":
        changed = rl_music_pause(this.handle) || changed;
        break;
      case "stop":
        changed = rl_music_stop(this.handle) || changed;
        break;
      default:
        break;
    }

    this.pendingAction = "none";
    return changed;
  }

  async destroy(): Promise<void> {
    if (this.handle == null) {
      return;
    }

    set_music_resource_manager(this.resourceManager);
    await rl_music_destroy(this.handle);
    this.handle = 0;
    this.pendingAction = "none";
    this.settingsDirty = false;
  }

  get loop(): boolean { return this._loop; }
  set loop(value: boolean) {
    if (this._loop === value) {
      return;
    }
    this._loop = value;
    this.settingsDirty = true;
  }

  get volume(): number { return this._volume; }
  set volume(value: number) {
    if (this._volume === value) {
      return;
    }
    this._volume = value;
    this.settingsDirty = true;
  }
}
